### PR TITLE
gnome.compile_resources: Add ld binary method

### PR DESCRIPTION
**Pull request by Ninja-Koala**
Instead of generating a c file that gets compiled,
directly create object file with ld.

See https://gitlab.gnome.org/GNOME/glib/issues/1489

### CODE REVIEW

**Comment by ignatenkobrain** [comment]
Flake8 detected 14 issues on 6fd7ddf8558c9aa969c427726ef7f98abeb9d13d
Visit https://sider.review/gh/19784232/pull_requests/4222 to review the issues.

**Comment by jpakkane** [comment]
Since this seems to be a performance optimization, has the impact been measured? What is the performance difference between the current version, the version in glib 2.60 and this version?

Also, has the 2.60 version been tested to work? The glib version in CI is too old to verify that.

**Comment by jpakkane** [comment]
There is also a potential bug here. When cross compiling and the `ld` command is not in the cross file (which is common since we don't use it for anything), find_program will return the system linker which would then fail (because it would produce an ELF file for the incorrect platform, probably).

File: mesonbuild/modules/gnome.py [hunk 1]
```diff
@@ -218,7 +263,77 @@ def compile_resources(self, state, args, kwargs):
             h_kwargs['install_dir'] = kwargs.get('install_dir',
                                                  state.environment.coredata.get_builtin_option('includedir'))
         target_h = GResourceHeaderTarget(args[0] + '_h', state.subdir, state.subproject, h_kwargs)
-        rv = [target_c, target_h]
+
+        if gresource_ld_binary:
```
Comment (jpakkane) [inline mesonbuild/modules/gnome.py]: Please turn this into its own function. The functionality is fairly isolated and this function is getting way too long already.

File: mesonbuild/modules/gnome.py [hunk 2]
```diff
@@ -218,7 +263,77 @@ def compile_resources(self, state, args, kwargs):
...
+            linkerscript_string += '{\n'
+            linkerscript_string += '  .gresource.' + c_name_no_underscores + ' : ALIGN(8)\n'
+            linkerscript_string += '  {\n'
+            linkerscript_string += '    ' + c_name + '_resource_data = _binary_' + symbol_name + '_start;\n'
+            linkerscript_string += '  }\n'
+            linkerscript_string += '  .data :\n'
+            linkerscript_string += '  {\n'
+            linkerscript_string += '    *(.data)\n'
+            linkerscript_string += '  }\n'
+            linkerscript_string += '}'
```
Comment (jpakkane) [inline mesonbuild/modules/gnome.py]: This should use a triple quoted string instead. It is more readable.

**Comment by Ninja-Koala** [comment]
Sorry for the long delay, didn't have time until recently

Regarding performance impact:
The impact is much higher for bigger files, for one 50MB resource
the current version needs 277.45s, glib 2.60 needs 67.44s and this variant needs 1.66s seconds.

When building 100 empty resources the current version needs around 51 seconds, glib 2.60 about 
53 seconds and this variant about 59 seconds. (All including build file generation)

I've done testing on my machine only, but the meson tests ran through and i also built some projects using my version of meson and glib succesfully.

**Pull request merged**